User Profile Backend

SimonFromSweden opened the pull request:
# User Profile Backend

## Description

This PR tries to develop the backend for the user profile page, so that users will be able to load their specific user info from the database and edit it as well. It adds two controller-functions in the userController that allows for getting and editing user profile data. This PR is tightly related to pr which develops the frontend of the user profile to be able to see, edit and save edits by clicking a button.

## Changes Made

List and describe the main changes in this PR.

- **API Endpoints:** Added the getUserProfile and updateUserProfile functions to userController. Both are located at "/api/user/profile" but one is has method GET and one has PUT
- **Business Logic:** userController have 2 new functions
- **Database:** No model/schema changes yet (though I tried extracting validateUser from "/api/user/profile"
- **Security:** Added requireAuth to the userRoutes created

## Testing Steps

There are still some steps to be done in order to make this backend update actually work with the frontend updates that are done in tandem with these.

I think this PR is testable by making postman requests to "localhost:3000/api/user/profile"

## Issues Addressed

This PR closes - [Task #133](https://github.com/orgs/Donna-Vino-Aps/projects/4/views/1?pane=issue&itemId=105319078&issue=Donna-Vino-Aps%7Cdonna-vino-ecommerce%7C133)

halynabondar reviewed (changes requested):
The backend currently relies on too many interdependent factors.

Authentication is being used as a data source for endpoints, which is not correct.

For example, the requireAuth middleware stores user data in the request object. This makes the middleware responsible for multiple things: both verifying authentication and preparing data for the API. This breaks the single responsibility principle. A middleware should only check if a user is authenticated and return a 401 if not — nothing more.

Additionally, the endpoint design doesn’t follow standard REST principles. A well-structured RESTful API should follow the CRUD pattern, which defines clear and predictable routes:

• GET /<entity>s – list all entities (e.g., GET /users) [List]
• POST /<entity> – create a new entity [Create]
• GET /<entity>/:id – get a specific entity [Retrieve]
• PUT /<entity>/:id – update a specific entity [Update]
• DELETE /<entity>/:id – delete a specific entity [Delete]

Following this pattern allows you to separate concerns — the authentication middleware (requireAuth) should only protect access to the route, while route handlers should deal with data logic.

For example, accessing an :id parameter in a route works natively:

```js
router.get('/path/to/:id', (req, res) => {
  const id = req.params.id; // <-- Access the `:id` here
  res.send(`The ID is ${id}`);
});
```

This design also means you could temporarily comment out requireAuth and the routes would still function — only access control would be disabled, not the core logic.

halynabondar commented on src/controllers/userController/userController.js:
```diff
@@ -12,3 +13,106 @@ export const getUsers = async (req, res) => {
+  }
+
+  try {
+    // Assuming you are storing the user's ID in the JWT token and using it to fetch the user
```
this comment is misleading now, as your id fetched from the request path

halynabondar commented on src/controllers/userController/userController.js:
```diff
@@ -12,3 +13,106 @@ export const getUsers = async (req, res) => {
+
+  try {
+    // Assuming you are storing the user's ID in the JWT token and using it to fetch the user
+    const user = await User.findById(id).select("-password"); // req.user.id comes from the 'protect' middleware (decoded JWT token)
```
Removing a password does not make any sense here, because:
 a) its encrypted
 b) you manually form a response object couple lines below, therefore you can simply not use that field

SimonFromSweden replied:
Ok I will remove this **select("-password")**

halynabondar commented on src/controllers/userController/userController.js:
```diff
@@ -12,3 +13,106 @@ export const getUsers = async (req, res) => {
+      });
+    }
+
+    if (email) {
```
Email should not be changed without any confirmation, this is crucial change!
Because email = login, and if you make a typo you will basically destroy the account.

This field must never be changed by this endpoint, for that should be a separate endpoint/form created with a mid-step: email confirmation.

SimonFromSweden replied:
Good point! Do you think the email-input should be removed from the user profile page? Or should we maybe include a "change email button" on the frontend leading to this separate form?

SimonFromSweden commented:
A few updates were made:

- I merged the branch with the latest changes from main
- I removed requireAuth from userRoutes
- I changed the authentication logic in userController to match the new authentication-system
- I removed the ability to change e-mail in the UpdateUserProfile-function in userController

halynabondar reviewed (changes requested):
Great work overall.
The implementation looks solid! I just added a few small comments around environment variable handling, route cleanup, and some code that’s no longer in use. Nothing major, mostly suggestions to keep things clean and consistent.

SimonFromSweden commented on src/config/environment.js:
```diff
@@ -39,17 +39,17 @@ if (!DONNA_VINO_ECOMMERCE_WEB_HEROKU || !DONNA_VINO_ECOMMERCE_WEB_LOCAL) {
 
 export const baseApiUrl =
   NODE_ENV === "production"
```
Fixed!

SimonFromSweden commented on src/routes/userRoutes.js:
```diff
@@ -1,8 +1,17 @@
+import {
+  getUsers,
+  getUserProfile,
+  updateUserProfile,
```
I will add a simple controller function for this as well as a route

SimonFromSweden commented on src/services/email/verificationEmailService.js:
```diff
@@ -9,7 +9,10 @@ export const sendVerificationEmail = async (user) => {
     const token = await generateToken(user.email);
 
-    const verifyUrl = `${baseApiUrl}/api/auth/sign-up?token=${token}`;
+    const verifyUrl = `${baseApiUrl.replace(
```
Yes, but that is done in the "refactor tokenGenerator PR", so I probably shouldnt touch it here right?

SimonFromSweden replied:
I reset the line to the old one

Natata08 reviewed (changes requested):
Simon, you are almost done with this task. I leave some comments to make them work and improve consistency. 
Also, we need to add a new validation for user profile updates. There are three approaches (at least):
1. Controller-Level Validation
How it works: Add validation checks directly in your controller function.
Something like:
```
if (firstName && (typeof firstName !== 'string' || firstName.length < 2)) {
  return res.status(400).json({ error: "Invalid first name" });
}
```
Good for: Small projects with few validation rules. 
Downside: Gets repetitive and messy in larger projects.

2. Middleware Validation Using Mongoose
How it works: Create a middleware that validates only the fields being updated.
In `userValidationMiddleware.js` we can create a temporary user object, validate only updated fields, and attache clean data to req.validatedUpdateData
Good for: Reusing your existing Mongoose schemas without duplicating rules.

3. Validation Libraries
How it works: Use libraries like express-validator, joi, or yup.
Good for: Complex validation with many rules. Benefits: Powerful features like conditional validation, custom error messages.

By implementing proper validation, we'll prevent bad data from entering our database and give users helpful error messages when something's wrong with their input.

SimonFromSweden commented on src/controllers/userController/userController.js:
```diff
@@ -1,5 +1,6 @@
 import { logError } from "../../util/logging.js";
+import { validateUser } from "../../models/users/userModels.js";
 
 export const getUsers = async (req, res) => {
```
I removed it!

SimonFromSweden commented on src/routes/userRoutes.js:
```diff
@@ -1,8 +1,17 @@
 const userRouter = express.Router();
 
 userRouter.get("/", getUsers);
+userRouter.get("/profile/:id", getUserProfile);
```
Done!

SimonFromSweden commented on src/controllers/userController/userController.js:
```diff
@@ -12,3 +13,122 @@ export const getUsers = async (req, res) => {
 };
+
+export const getUserProfile = async (req, res) => {
+  const userIdFromToken = req.accessToken.userId;
```
Nice catch!

SimonFromSweden commented on src/routes/userRoutes.js:
```diff
@@ -1,8 +1,17 @@
 
 userRouter.get("/", getUsers);
+userRouter.get("/profile/:id", getUserProfile);
+userRouter.patch("/profile/:id", updateUserProfile);
```
Meaning I will leave the routes like this?

userRouter.get("/:id", authMiddleware, getUserProfile);
userRouter.patch("/:id", authMiddleware, updateUserProfile);
userRouter.delete("/:id", authMiddleware, deleteUser);

SimonFromSweden commented on src/controllers/userController/userController.js:
```diff
@@ -1,5 +1,6 @@
 import User from "../../models/userModels.js";
 import { logError } from "../../util/logging.js";
+import { validateUser } from "../../models/users/userModels.js";
```
Done

SimonFromSweden commented:
> Simon, you are almost done with this task. I left some comments to make them work and improve consistency. Also, we need to add a new validation for user profile updates. There are three approaches (at least):
> 
> 1. Controller-Level Validation
>    How it works: Add validation checks directly in your controller function.
>    Something like:
> 
> ```
> if (firstName && (typeof firstName !== 'string' || firstName.length < 2)) {
>   return res.status(400).json({ error: "Invalid first name" });
> }
> ```
> 
> Good for: Small projects with few validation rules. Downside: Gets repetitive and messy in larger projects.
> 
> 2. Middleware Validation Using Mongoose
>    How it works: Create a middleware that validates only the fields being updated.
>    In `userValidationMiddleware.js` we can create a temporary user object, validate only updated fields, and attache clean data to req.validatedUpdateData
>    Good for: Reusing our existing Mongoose schemas without duplicating rules.
> 3. Validation Libraries
>    How it works: Use libraries like express-validator, joi, or yup.
>    Good for: Complex validation with many rules. Benefits: Powerful features like conditional validation, custom error messages.
> 
> By implementing proper validation, we'll prevent bad data from entering our database and give users helpful error messages when something's wrong with their input.

I tried achieving this in the new userValidationMiddleware

Natata08 reviewed (approved):
Works perfectly and ready to be merged!
